src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -103,12 +119,154 @@ void GcsActorManager::RegisterActor(
 
   auto actor = std::make_shared<GcsActor>(request);
   RAY_CHECK(registered_actors_.emplace(actor->GetActorID(), actor).second);
+
+  if (!actor->IsDetached() && worker_client_factory_) {
+    // This actor is owned. Send a long polling request to the actor's
+    // owner to determine when the actor should be removed.
+    PollOwnerForActorOutOfScope(actor);
+  }
+
   gcs_actor_scheduler_->Schedule(actor);
 }
 
-void GcsActorManager::ReconstructActorOnWorker(const ray::ClientID &node_id,
-                                               const ray::WorkerID &worker_id,
-                                               bool need_reschedule) {
+void GcsActorManager::PollOwnerForActorOutOfScope(
+    const std::shared_ptr<GcsActor> &actor) {
+  const auto &actor_id = actor->GetActorID();
+  const auto &owner_node_id = actor->GetOwnerNodeID();
+  const auto &owner_id = actor->GetOwnerID();
+  auto &workers = owners_[owner_node_id];
+  auto it = workers.find(owner_id);
+  if (it == workers.end()) {
+    RAY_LOG(DEBUG) << "Adding owner " << owner_id << " of actor " << actor_id;
+    std::shared_ptr<rpc::CoreWorkerClientInterface> client =
+        worker_client_factory_(actor->GetOwnerAddress());
+    it = workers.emplace(owner_id, Owner(std::move(client))).first;
+  }
+  it->second.children_actor_ids.insert(actor_id);
+
+  rpc::WaitForActorOutOfScopeRequest wait_request;
+  wait_request.set_intended_worker_id(owner_id.Binary());
+  wait_request.set_actor_id(actor_id.Binary());
+  RAY_CHECK_OK(it->second.client->WaitForActorOutOfScope(
+      wait_request, [this, owner_node_id, owner_id, actor_id](
+                        Status status, const rpc::WaitForActorOutOfScopeReply &reply) {
+        if (!status.ok()) {
+          RAY_LOG(INFO) << "Worker " << owner_id << " failed, destroying actor child";
+        }
+
+        auto node_it = owners_.find(owner_node_id);
+        if (node_it != owners_.end() && node_it->second.count(owner_id)) {
+          // Only destroy the actor if its owner is still alive. The actor may
+          // have already been destroyed if the owner died.
+          DestroyActor(actor_id);
+        }
+      }));
+}
+
+void GcsActorManager::DestroyActor(const ActorID &actor_id) {
```

Comment:
We'll need to implement a "push-based" actor destroy handler for detached actors as well. Will it just be as simple as calling this method?

Reply:
Yep!